ruby/scan_kwoptargs/scan_keyword_arguments.rb

```diff
@@ -0,0 +1,162 @@
+#
+# But as this is just one time script it is easy to manually check the results.
+#
+# Links:
+# - https://docs.rubocop.org/rubocop-ast/index.html
+# - https://docs.rubocop.org/rubocop-ast/node_types.html
+#
+# This script runs two passes, in the first one it collects the method definitions,
+# in the second one it tries to match all method calls with method definitions
+# and find the difference in the arguments.
+#
+# You can use "ruby-parse" tool to see the parsed AST for a specific Ruby code,
+# e.g. "ruby-parse -e 'foo(bar: baz)'".
+#
+# Testing:
+#  mkdir test
+#  curl https://raw.githubusercontent.com/yast/yast-yast2/c9127aad823c9d909f58940745ba981547c5f37f/library/cwm/src/modules/CWM.rb > test/CWM.rb
+#  curl https://raw.githubusercontent.com/yast/yast-iscsi-client/fe946fdb9a12dda6fb062786129dbbed99ca5263/src/include/iscsi-client/dialogs.rb > test/dialogs.rb
+#  ./scan_keyword_arguments.rb test
+
+# install missing gems
+unless File.exist?("./.vendor")
+  puts "Installing the needed Rubygems to ./.vendor/bundle ..."
+  system "bundle install --path .vendor/bundle"
+end
+
+require "rubygems"
+require "bundler/setup"
+
+require "rubocop-ast"
+
+# helper method to find the parent class node
+def find_parent_class_name(node)
+  parent_node = node.parent
+  # no parent
+  return nil if parent_node.nil?
+
+  # found a class
+  if parent_node.class_type?
+    # only constant class name
+    return parent_node.identifier.short_name if parent_node.identifier.const_type?
+
+    return nil
+  end
+
+  # search recursively
+  find_parent_class_name(parent_node)
+end
+
+# a human readable node location in the source file (in "file:line" format)
+def node_location(node)
+  "#{node.location.expression.source_buffer.name}:#{node.location.line}"
+end
+
+# Collect all methods which use keyword arguments
+class MethodCollector < Parser::AST::Processor
+  include RuboCop::AST::Traversal
+
+  attr_reader :kwoptarg_methods
+
+  # TODO: maybe use a node pattern matching here?
+  # https://docs.rubocop.org/rubocop-ast/node_pattern.html
+  # extend RuboCop::AST::NodePattern::Macros
+  # def_node_matcher :method_with_kwoptarg?, '(def ?????)'
+
+  def initialize
+    @kwoptarg_methods = Set.new
+    super
+  end
+
+  def on_def(node)
+    if node.arguments.any?(&:kwoptarg_type?) && node.arguments.any?(&:arg_type?) && find_parent_class_name(node)
+      # puts "Found method with kwoptargs #{node.method_name.to_s.inspect} at " \
+      #    "#{node.location.expression.source_buffer.name}:#{node.location.line}"
+      kwoptarg_methods << node
+    end
+  end
+end
+
+# Check if the called method matches the definition
+class CallChecker < Parser::AST::Processor
+  include RuboCop::AST::Traversal
+
+  attr_reader :method_defs
+
+  def initialize(definitions)
+    @method_defs = definitions
+    super()
+  end
+
+  def on_send(node)
+    # none receiver
+    return if node.receiver.nil? || !node.receiver.const_type?
```

Comment:
The `const_type?` is a crucial part of why this works: because the receiver is a YaST singleton `Foo` of the type `FooClass`.
I'd bet it covers 90% of our call sites, but we should stop and think what APIs this might be missing. Recent APIs in storage-ng, registration, network?
Would it be worth it to make a more fuzzy run in those modules known to use non-singleton APIs?
Maybe: if the `send` does not match a singleton `def`, and it uses keyword args, print it

Reply:
> Maybe: if the `send` does not match a singleton `def`, and it uses keyword args, print it

I tried that and I got several thousands results, that's too much for a manual check. We would need to add some additional logic or at least ignore some most common false positives.

But I do not think we should invest more time here, it already took me quite a lot of time and I do not expect that this kind of problem happens often in the YaST code base...

Reply:
OK then, this is enough.